python/cuml/cuml/common/classification.py

```diff
@@ -32,6 +33,198 @@ def check_classification_targets(y):
+        y_encoded = cp.empty(shape=y.shape, dtype=dtype, order=order)
+        for i, col in enumerate(encoded_cols):
+            y_encoded[:, i] = col
+
+    return y_encoded, classes
+
+
+def decode_labels(y_encoded, classes, output_type="cupy"):
+    """Convert encoded labels back into their original classes.
+
+    Parameters
+    ----------
+    y_encoded : cp.ndarray
+        The labels, encoded as integers in [0, n_classes - 1].
+    classes : np.ndarray or list[np.ndarray]
+        The array of classes, or a list of arrays if multi-target.
+    output_type : str, optional
+        The type to output. May be any of the output types cuml supports.
+
+    Returns
+    -------
+    labels
+        The decoded labels, as output type ``output_type``.
+    """
+    if isinstance(classes, list):
+        # Multi-target output
+        dtype = (
+            classes[0].dtype
+            if len(set(c.dtype for c in classes)) == 1
+            else None
+        )
+        if dtype is not None and dtype.kind in "iufb":
+            # All dtypes are identical and numeric, we can use cupy here
+            if all((c == np.arange(len(c))).all() for c in classes):
+                # Fast path for common case of monotonically increasing numeric classes
+                labels = y_encoded.astype(dtype, copy=False)
+            else:
+                # Need to transform y_encoded back to classes
+                labels = cp.empty(shape=y_encoded.shape, dtype=dtype)
+                for i, c in enumerate(classes):
+                    labels[:, i] = cp.asarray(c).take(y_encoded[:, i])
+
+            out = CumlArray(labels)
+        else:
+            # At least one class is non-numeric, we need to use cudf
+            out = cudf.DataFrame(
+                {
+                    i: cudf.Series(c)
+                    .take(y_encoded[:, i])
+                    .reset_index(drop=True)
+                    for i, c in enumerate(classes)
+                }
+            )
+    else:
+        # Single-target output
+        dtype = classes.dtype
+        if classes.dtype.kind in "iufb":
+            # Numeric dtype, we can use cupy here
+            if (classes == np.arange(len(classes))).all():
+                # Fast path for common case of monotonically increasing numeric classes
+                labels = y_encoded.astype(classes.dtype, copy=False)
+            else:
+                # Need to transform y_encoded back to classes
+                labels = cp.asarray(classes).take(y_encoded)
+
+            out = CumlArray(labels)
+        else:
+            # Non-numeric classes. We use cudf since it supports all types, and will
+            # error appropriately later on when converting to outputs like `cupy`
+            # that don't support strings.
+            out = cudf.Series(classes).take(y_encoded).reset_index(drop=True)
+
+    # Coerce result to requested output_type
+    if isinstance(out, CumlArray):
```

Comment:
This is messy, but is basically a generalization of what we already added to `LogisticRegression` months ago. I hope we can simplify this a bunch when we cleanup our output type handling.